### PR TITLE
Crypto: It should follow the stream design pattern.

### PR DESCRIPTION
**Pull request by cabelitos**
Signed-off-by: Guilherme Iscaro <guilherme.iscaro@intel.com>

### CODE REVIEW

File: src/lib/crypto/include/sol-message-digest.h [hunk 1]
```diff
@@ -151,15 +151,27 @@ struct sol_message_digest_config {
...
+     * The tx buffer max size. The value @c 0 means unlimited data.
+     * Since sol_message_digest_feed() works with blobs, no extra buffers will be allocated in order
+     * to store @c tx_size bytes. All the blobs that are schedule to be written will be referenced
+     * and the sum of all queued blobs must not be equal or exceed @c tx_size.
+     * If it happens sol_message_digest_feed() will return @c -ENOSPC and one must start to control the
+     * writing flow until @c on_feed_done is called.
+     * @see sol_message_digest_feed()
+     */
+    size_t tx_size;
 };
```
Comment (barbieri) [inline src/lib/crypto/include/sol-message-digest.h]: you should follow uart and mention the pattern in the sol_message_digest_config structure.

I'm unsure if we rename `tx_size` to `feed_size` or `on_feed_done` to `tx_cb`, ask others to chime in

File: src/lib/crypto/include/sol-message-digest.h [hunk 2]
```diff
@@ -218,8 +230,8 @@ void sol_message_digest_del(struct sol_message_digest *handle);
  *        @c config->on_digest_ready() is called with the resulting digest
  *        as a blob.
  *
- * @return 0 on success, -errno otherwise. If error, then the input reference
- *         is not taken.
+ * @return 0 on success, @c -ENOSPC if sol_message_digest_config::tx_size is not zero and there's
+ * no more space left or -errno on error. If error or -ENOPSC, If error, then the input reference is not taken.
```
Comment (barbieri) [inline src/lib/crypto/include/sol-message-digest.h]: link to the design pattern page, as done with UART

File: src/lib/crypto/include/sol-message-digest.h (hunk 1 above)
Comment (barbieri) [inline src/lib/crypto/include/sol-message-digest.h]: tx will be strange, keep it `feed`, but maybe mention this is an alias in the design pattern.

**Comment by barbieri** [comment]
review the code comparing to your written design pattern. Maybe we'll need to change the design pattern to mention 'feed' instead of `tx`, so it may be applicable in more scenarios.

Also note that the `on_feed_done()` doesn't respect the "delete from callback", we'll need the delete_me/in_use booleans to do that.

File: src/lib/crypto/include/sol-message-digest.h (hunk 2 above)
Comment (cabelitos) [inline src/lib/crypto/include/sol-message-digest.h]: I did not mention because the docs are not part of this PR. I don't want to break the docs generation

**Comment by cabelitos** [comment]
@barbieri I did not worry about the delete_me/in_use because the docs says ` * @note it is safe to delete the message digest handle from  inside this callback. `

File: src/lib/crypto/include/sol-message-digest.h (hunk 1 above)
Comment (cabelitos) [inline src/lib/crypto/include/sol-message-digest.h]: To be honest I don't like `tx_....` prefix. I guess we should change it to `feed`.

**Comment by barbieri** [comment]
ah, that's right, there is a refcnt in the handle and it's increased before calling the user.

**Comment by barbieri** [comment]
btw, to follow your new design pattern, the _sol_message_digest_free() must be changed to call `on_feed_done(... , -ECANCELLED)` for pending blobs.

**Comment by cabelitos** [comment]
@barbieri It's calling with `-ECANCELLED`

File: src/lib/crypto/sol-message-digest-common.c [hunk 3]
```diff
@@ -317,6 +320,8 @@ _sol_message_digest_free(struct sol_message_digest *handle)
     _sol_message_digest_thread_fini(handle);
 
     SOL_VECTOR_FOREACH_IDX (&handle->pending_feed, pf, i) {
+        if (handle->on_feed_done)
+            handle->on_feed_done((void *)handle->data, handle, pf->blob, -ECANCELED);
```
Comment (cabelitos) [inline src/lib/crypto/sol-message-digest-common.c]: Right here

Comment (barbieri) [inline src/lib/crypto/sol-message-digest-common.c]: hum... but that just accounts what is to be processed. Look into `_sol_message_digest_thread_fini()`, what was processed but not dispatched, will still be in `handle->pending_dispatch`.

This list is consumed from `_sol_message_digest_thread_feedback()`, where the vector is stolen to a local variable and re-initialized to zero (took me few minutes to get that, and I was the one who wrote the code... ouch!)

Comment (cabelitos) [inline src/lib/crypto/sol-message-digest-common.c]: You're right. I will add that

**Comment by cabelitos** [comment]
#2048
